### PR TITLE
Improve console output

### PR DESCRIPTION
**Pull request by PhilippMatthes**
Fixes: #236 

Hi @martinmo, I removed the erroneous whitespacing and visually combined the `stderr` with the `stdout` of a TestResult in the console output view. 

However, I don't know exactly how much it would affect the stability of the deployed system if we'd remove the `stderr` attribute on the `TestResult` model. So what I simply did is to pipe the stderr output of a testrun to `stdout`:

```python
proc = subprocess.Popen(
    args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT
)
```

This interleaves the `stderr` output with the `stdout` output. However, there will be no more `stderr`.

@martinmo so my question is, can I remove the `stderr` attribute of `TestResult` and how would that affect our deployed system? With that done, I could simply return the output instead of both `stderr` and `stdout` separately and store it into the `stdout` attribute all at once. Of course, this would make a template change necessary. Implementation-wise, this should all be no problem since I already made migrations before, but I wanted to simply ask before performing potential risky changes.

### CODE REVIEW

**Comment by PhilippMatthes** [comment]
There are currently 2 tests failing, because the output is now interleaved. I'll fix this later when it's clear where we want to go with this.

**Comment by martinmo** [comment]
> so my question is, can I remove the stderr attribute of TestResult and how would that affect our deployed system?

This is a good question! If we remove the field, all the solutions that have been submitted until now would be changed.

I would say, we keep both stderr and stdout fields in the model, but pipe stderr to stdout as you did. At a later point we could even make this configurable. You'll need to change the tests for the TestRunner class accordingly.

File: inloop/testrunner/runner.py [hunk 1]
```diff
@@ -160,8 +161,14 @@ def communicate(self, task_name, input_path, output_path):
             LOG.debug("removing timed out container %s", ctr_id)
             subprocess.call(["docker", "rm", "--force", str(ctr_id)], stdout=subprocess.DEVNULL)
 
-        stderr = stderr.decode("utf-8", errors="replace")
-        stdout = stdout.decode("utf-8", errors="replace")
+        if stderr:
```
Comment (martinmo) [inline inloop/testrunner/runner.py]: Out of curiosity: does piping to `stdout` mean `stderr` will be `None`? Or why is this check necessary? `"".decode()` should work afaik.

**Comment by PhilippMatthes** [comment]
Hi again, yes, after interleaving both streams, stderr is always None. I could also change the method signature so that it doesn't return stderr anymore. But this would result in a cascade of code changes, ending at the point where I create the TestResult object. So I just defaulted stderr to "".

**Comment by PhilippMatthes** [comment]
@martinmo I tried to rewrite the tests but they are failing once in two tries. The reason for this is that 
```shell
echo -n OUT; echo -n ERR >&2; echo -n OUT;
```
is either written to `stdout` as
```shell
OUTOUTERR
```
or
```shell
ERROUTOUT
```
which indicates that the pipes are not interleaved correctly. However, when I do this in my python shell:
```python
>>> import subprocess
>>> cmd = "echo -n OUT; echo -n ERR >&2; echo -n OUT; exit 42"
>>> proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True)
>>> stdout, stderr = proc.communicate()
>>> stdout
b'-n OUT\n-n ERR\n-n OUT\n'
>>> 
```
...both outputs are deterministically and correctly interleaved. 

I find this really odd, maybe it is due to a race condition in Docker. Do you have any idea, why Docker doesn't interleave both streams correctly?

**Comment by PhilippMatthes** [comment]
I added a new test to validate, if both outputs are interleaved correctly. As discussed, this test is currently failing since it expects `OUTERROUT` but receives either `OUTOUTERR` or `ERROUTOUT`.

**Comment by PhilippMatthes** [comment]
I don't get why *travis* is failing on python 3.4 with 
```shell
error parsing HTTP 404 response body: invalid character 'p' after top-level value: "404 page not found\n"
```
Hopefully, this is only a temporary error.

**Comment by PhilippMatthes** [comment]
Wow, thank you for the detailed explanation! I will propagate your proposed changes and seek back to you.

**Comment by PhilippMatthes** [comment]
Removing the `in progress` label, because this issue is a bit more complicated than we initially thought. We'll come back to this later when more important tasks are finished.

**Pull request closed without merging**